activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -64,6 +66,27 @@ def find_finder_class_for(record)
         class_hierarchy.detect { |klass| !klass.abstract_class? }
       end
 
+      def unchanged_and_covered_by_unique_index?(klass, record, attribute)
+        return false if options[:conditions] || options.key?(:case_sensitive)
+
+        scope = Array(options[:scope]).map(&:to_s)
+        return false unless attributes_scope?(record, scope)
+
+        attributes = scope + [attribute.to_s]
+        return false if attributes.any? { |attribute| record.attribute_changed?(attribute) }
+
+        klass.connection.schema_cache.indexes(klass.table_name).any? do |index|
+          index.unique &&
+            index.where.nil? &&
+            index.columns.size <= attributes.size &&
+            attributes[0, index.columns.size] == index.columns
+        end
+      end
```

Comment:
A lot of things in here could be computed just once, as it only depends on the schema and the validator options.

The problem is we can't access the schema when the model is first defined, so we'll have to memoize on the first call...

Reply:
Do you mean memoization in validator's instance variable? Is this thread-safe?

Reply:
Well. Thread-safe might mean a lot of different things.

It is indeed subject to a race-condition, but since we're computing a value based on static data, if two threads enter the code path, all it means is that we're a little bit wasteful because we compute the thing twice, but in the end both end up with the same result.

That's for MRI "thanks" to the GVL. Now I'm not certain about JRuby / Truffle, I know they sometimes raise if you access some datastructures concurrently, but I don't remember if it applies to instance variables. https://github.com/jruby/jruby/wiki/Concurrency-in-jruby suggest it's fine.